### PR TITLE
DDF-3836 Extend query abilities to support multiple geometries

### PR DESCRIPTION
**Pull request by lamhuy**
#### What does this PR do?
OpenSearch capabilities (Search/Endpoint) to OpenSearch geo Draft 2 by allowing geometries instead of polygons

OpenSearch endpoints are both upgraded to accept geometries and polygons.

OpenSearch endpoint support multiple spatial criteria at the same time (geometry, polygon, bounding box, point radius).

OpenSearch Sources generate a GeometryCollection to put in the geometry field when user defines multiple geospatial criteria in the UI.

OpenSearch Sources generate a polygon to put in the geometry field when user defines a polygon geospatial criteria in the UI.

With convertToBoundingBox as false, if OpenSearch Source encounters multiple point radius search, an approximate polygon (square) is formulated and add to a geometry collection.

With convertToBoundingBox as true, if OpenSearch Source encounters multiple point radius search, an an approximate polygon (square) is formulated and add to a geometry collection. Then the bounding box is computed from the geometry collection's envelope encompassing all the geometries. Resulting in a wider search

DDF loopback federation correctly supports multiple geometries. 

DDF loopback federation correctly supports polygon searches. 

#### Who is reviewing it? 
@millerw8 
@bdeining 
@mackncheesiest 
@mojogitoverhere 

#### Select relevant component teams: 
https://github.com/orgs/codice/teams

#### Ask 2 committers to review/merge the PR and tag them here. If you don't know who to ask, you can request reviews in https://groups.google.com/forum/#!forum/ddf-developers .
@bdeining
@millerw8

#### How should this be tested? (List steps with links to updated documentation)
Setup:
1. On karaf console: log:set INFO org.apache.cxf
2. Configure a Opensearch source. Turn off Convert to Bounding Box
3. Upload a geotagged image onto DDF
4. Open Advanced search dialog

Case 1
Specify a polygon and a bounding box, with one containing the uploaded image
Verify record is found. OR operation the spatial criteria. As they are combined to a Geometry Collection
Check DDF cxf log for OpenSearch request url. Verify geometry argument containing a GeometryCollection with 2 polygons

Case 2
Specify a polygon and a point radius with one containing the uploaded image
Verify the record is NOT found.  AND operation between the spatial criteria
https://codice.atlassian.net/browse/DDF-3857 to address this issue
Check DDF cxf log for OpenSearch request url. Verify a geometry argument as a POLYGON and a lat/long/radius argument.

Case 3
Specify two point radius with one containing the uploaded image or at the edge of the point radius
Verify the record is found.  Point radius is converted into two approximate polygon (square) that's group as a Geometry Collection
Check DDF cxf log for OpenSearch request url. Verify a GeometryCollection with 2 polygons

Case 4
Turn on convertToBoudningBox = true
Specify a polygon and a point radius with the uploaded image between the two spatial criteria
Verify the record is found.
Check DDF cxf log verify a large bounding box envelope was used for the search that encompassed the two spatial criteria and the result. 

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-3836

#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java [hunk 1]
```diff
@@ -668,6 +685,67 @@ public void testPolygonSpatialFilter() {
...
+        "30.943,-120.032,35.039,-120.032,35.039,-110.856,30.943,-110.856,30.943,-120.032");
+    query.addBBoxSpatialFilter("-120.032,30.943,-110.856,35.039");
+    query.addPointRadiusSpatialFilter("117.3425", "33.9283", "5000");
+    query.addPolygonSpatialFilter(
+        "30.943,-120.032,35.039,-120.032,35.039,-110.856,30.943,-110.856,30.943,-120.032");
+
+    Filter filter = query.getFilter();
+    assertThat(filter, notNullValue());
+
+    // TODO: fix as OR filter after https://jira.di2e.net/browse/GSR-2130
```
Comment (millerw8) [inline catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: make ddf story for this

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/endpoint/OpenSearchEndpoint.java [hunk 2]
```diff
@@ -276,13 +276,19 @@ private void addSpatialFilter(
     if (StringUtils.isNotBlank(geometry)) {
       LOGGER.trace("Adding SpatialCriterion geometry: {}", geometry);
       query.addGeometrySpatialFilter(geometry.trim());
-    } else if (StringUtils.isNotBlank(bbox)) {
+    }
+
+    if (StringUtils.isNotBlank(bbox)) {
```
Comment (leo-sakh) [inline catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/endpoint/OpenSearchEndpoint.java]: This whole method 'feels' odd. Passing in the 'query' with parameters to be added in a void method seems 'side-effecty'. I would suggest refactoring this to at least return the query, but better to create one via a copy constructor or even a builder.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 3]
```diff
@@ -725,7 +728,7 @@ protected InputTransformer lookupTransformerReference(String namespaceUri)
   }
 
   @VisibleForTesting
-  Bundle getBundle() {
+  protected Bundle getBundle() {
     return FrameworkUtil.getBundle(this.getClass());
```
Comment (leo-sakh) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: I would strongly suggest to actually supply the service references you need in the constructor instead of using the static `FrameworkUtil` as its hard to test (as evidenced by the annotation on the method) and it hides what you actually need.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 4]
```diff
@@ -1043,50 +1054,45 @@ public PointRadius getPointRadius() {
   @Nullable
   protected SpatialSearch createCombinedSpatialSearch(
```
Comment (leo-sakh) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: I would suggest splitting this method into two, one dealing with point radius searches and the other with geometry searches. As it is, the operations performed on each are distinct and do not appear to depend on each other. This will let you get rid of the arrow if/else anti-pattern and since the current method is only called in one place would be minimally disruptive.

Comment (leo-sakh) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: Also this class is massive, consider refactoring it.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 5]
```diff
@@ -1043,50 +1054,45 @@ public PointRadius getPointRadius() {
...
-          ? new SpatialSearch(BoundingBoxUtils.createBoundingBox(polygon))
-          : new SpatialSearch(polygon);
     }
 
-    if (filterContainedSomePointRadiusCriteria) {
-      if (CollectionUtils.size(pointRadiusSearches) > 1) {
-        LOGGER.debug(
-            "Ignoring all spatial searches because combining point radius-searches is not yet implemented.");
-        return null;
+    if (hasPointRadiusSearch) {
```
Comment (millerw8) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: Should at least always be able to convert to a bbox envelope for all geospatial constraints.  In the case of multiple point/radius without converting to bbox overall, could convert the individual point radius to their envelope and add them as polygons to the geometry collection to have an approximation

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java (hunk 4 above)
Comment (bdeining) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: while this is good feedback, these comments should be considered ✏️

File: catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/endpoint/OpenSearchEndpoint.java (hunk 2 above)
Comment (bdeining) [reply catalog/opensearch/catalog-opensearch-endpoint/src/main/java/org/codice/ddf/opensearch/endpoint/OpenSearchEndpoint.java]: while this is good feedback, these comments should be considered ✏️

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java (hunk 5 above)
Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: Yes, making update to this logic shortly

File: catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java (hunk 1 above)
Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: updated https://codice.atlassian.net/browse/DDF-3857

**Comment by lamhuy** [comment]
build now

File: catalog/opensearch/catalog-opensearch-source/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 6]
```diff
@@ -73,6 +73,7 @@
                     <value>lon</value>
                     <value>radius</value>
                     <value>bbox</value>
+                    <value>geometry</value>
```
Comment (emmberk) [inline catalog/opensearch/catalog-opensearch-source/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: ❗️ The documentation also needs to be updated: https://github.com/codice/ddf/blob/43d0aff293e1a9c2f60447cd409013654f84540a/distribution/docs/src/main/resources/content/_tables/OpenSearchSource.adoc#L47-L52

File: catalog/opensearch/catalog-opensearch-source/pom.xml [hunk 7]
```diff
@@ -108,6 +108,10 @@
             <version>2.0.4</version>
             <scope>test</scope>
         </dependency>
+      <dependency>
+        <groupId>ddf.catalog.common</groupId>
+        <artifactId>geo-formatter</artifactId>
+      </dependency>
```
Comment (emmberk) [inline catalog/opensearch/catalog-opensearch-source/pom.xml]: ✏️ `<version>${project.version}</version>`

File: catalog/common/catalog-common-geo-formatter/src/main/java/ddf/geo/formatter/CompositeGeometry.java [hunk 8]
```diff
@@ -159,7 +159,7 @@ private static double getDouble(Object object) {
    * @return an array of {@link Coordinate} objects, if input is <code>null</code> or empty, it will
    *     return an empty array
    */
-  protected static Coordinate[] getCoordinates(List<List<String>> primitiveCoordinatesList) {
```
Comment (emmberk) [inline catalog/common/catalog-common-geo-formatter/src/main/java/ddf/geo/formatter/CompositeGeometry.java]: ✏️ Why was this changed from `List<List<String>>` to `List<List>`? Can we get away with specifying the `List` type?

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 9]
```diff
@@ -1044,50 +1056,48 @@ public PointRadius getPointRadius() {
   @Nullable
   protected SpatialSearch createCombinedSpatialSearch(
```
Comment (emmberk) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: ❗️ The java doc for this method needs to be updated.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java [hunk 10]
```diff
@@ -1044,50 +1056,48 @@ public PointRadius getPointRadius() {
...
-    if (filterContainedSomePointRadiusCriteria && filterContainedSomePolygonSearchCriteria) {
-      LOGGER.debug(
-          "Ignoring all spatial searches because combining spatial and point radius searches is not yet implemented.");
-      return null;
+    Geometry geometrySearch = null;
+    BoundingBox boundingBox = null;
+    PointRadius pointRadius = null;
+    SpatialSearch spatialSearch = null;
+
+    Queue<Geometry> combinedGeometrySearches = new LinkedList<>(geometrySearches);
```
Comment (emmberk) [inline catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: ❓ Why is this collection of `Geometry`s a `Queue`? Why not a `Set`?

File: catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java [hunk 11]
```diff
@@ -668,6 +685,67 @@ public void testPolygonSpatialFilter() {
...
+  @Test
+  public void testMultipleSpatialFilter() {
+    OpenSearchQuery query = new OpenSearchQuery(0, 10, "relevance", "desc", 30000, FILTER_BUILDER);
+    query.addGeometrySpatialFilter(GEOMETRY_WKT);
+    query.addPolygonSpatialFilter(
+        "30.943,-120.032,35.039,-120.032,35.039,-110.856,30.943,-110.856,30.943,-120.032");
+    query.addBBoxSpatialFilter("-120.032,30.943,-110.856,35.039");
+    query.addPointRadiusSpatialFilter("117.3425", "33.9283", "5000");
+    query.addPolygonSpatialFilter(
+        "30.943,-120.032,35.039,-120.032,35.039,-110.856,30.943,-110.856,30.943,-120.032");
```
[the remark this answers is not in the document]
Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: no, i mean to add multiple polygon criteria. i'll use a different polygon coordinate.

File: catalog/common/catalog-common-geo-formatter/src/main/java/ddf/geo/formatter/CompositeGeometry.java (hunk 8 above)
Comment (lamhuy) [reply catalog/common/catalog-common-geo-formatter/src/main/java/ddf/geo/formatter/CompositeGeometry.java]: i think so. This method can accept any Object, the Object will be converted to a double.

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java (hunk 10 above)
Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchSource.java]: changing to a set

File: catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchFilterVisitor.java [hunk 12]
```diff
@@ -483,8 +485,11 @@ protected static void buildGeometrySearch(BinarySpatialOperator filter, Object d
     } else if (geometryExpression instanceof Polygon) {
       Polygon polygon = (Polygon) literalWrapper.evaluate(null);
       openSearchFilterVisitorObject.addGeometrySearch(polygon);
+    } else if (geometryExpression instanceof GeometryImpl) {
+      Geometry polygon = ((GeometryImpl) geometryExpression).getJTSGeometry();
+      openSearchFilterVisitorObject.addGeometrySearch(polygon);
```
[the remark this answers is not in the document]
Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchFilterVisitor.java]: good `Point` GeometryImpl was added to support `Geometry Collection`, `Multipolygon` and apperantly `Point` as well.   So the two `*WithPoint` tests used hit the else LOGGER.debug block, it is not actually returning `geometrySearch` criteria. Will update existing tests and add additional tests

File: catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java [hunk 13]
```diff
@@ -66,6 +77,45 @@
...
+  private static final String POLYGON_WKT =
+      "POLYGON ((-120.032 30.943, -120.032 35.039, -110.856 35.039, -110.856 30.943, -120.032 30.943))";
+
+  private static final String POLYGON_WKT_2 =
+      "POLYGON ((100 -30, 100 -35, 110 -35, 110 -30, 100 -30))";
+
+  private static final String POINT_WKT = "POINT (117.3425 33.9283)";
+
+  private static final DWithin D_WITHIN_FILTER =
+      (DWithin)
```
Comment (bdeining) [inline catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: Do these need to be casted?

File: catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java [hunk 14]
```diff
@@ -668,6 +718,74 @@ public void testPolygonSpatialFilter() {
...
+  public void testGeometrySpatialFilter() {
+    OpenSearchQuery query = new OpenSearchQuery(0, 10, "relevance", "desc", 30000, FILTER_BUILDER);
+    query.addGeometrySpatialFilter(GEOMETRY_WKT);
+    Filter filter = query.getFilter();
+    assertThat(filter, notNullValue());
+    Intersects intersects = (Intersects) filter;
+    Literal literalWrapper = (Literal) intersects.getExpression2();
+    Object geometryExpression = literalWrapper.getValue();
+    assertThat(geometryExpression, instanceOf(GeometryImpl.class));
+    com.vividsolutions.jts.geom.Geometry polygon =
```
Comment (bdeining) [inline catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: do we need the full package name?

Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: yes, the default import is org.opengis.geometry.Geometry

File: catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java (hunk 13 above)
Comment (lamhuy) [reply catalog/opensearch/catalog-opensearch-endpoint/src/test/java/org/codice/ddf/opensearch/endpoint/query/OpenSearchQueryTest.java]: ye, compiler doesn't like it without.

**Comment by lamhuy** [comment]
build now

File: distribution/docs/src/main/resources/content/_sources/opensearch-source.adoc [hunk 15]
```diff
@@ -88,15 +89,21 @@ Use the OpenSearch source if querying a CDR-compliant search service is desired.
...
 //TODO DDF-3742
  +
-Or else, calculated from query if it is a point-radius query, the radius of point-radius query is > 0, and the <<shouldConvertToBBox,`shouldConvertToBBox`>> configuration option is `true`.
+Or else, calculated from the query if it is a geometry collection and the <<shouldConvertToBBox,`shouldConvertToBBox`>> configuration option is `true`.
+Note: An approximate bounding box is used to represent the geometry collection encompassing all of the geometries within it +
+Area between the geometries are also included in the bounding box. Hence widen the search area.
+
+|geometry
+|geometry
+|Pulled from the ${branding} query and combine as a geometry collection if multiple spatial query exist.
```
Comment (emmberk) [inline distribution/docs/src/main/resources/content/_sources/opensearch-source.adoc]: ✏️ combined

**Comment by lamhuy** [comment]
build now

**Comment by BernardIgiri** [comment]
Bounding box searching still fails for multiple search geos see [this video.](https://codice.atlassian.net/secure/attachment/18257/18257_multiple-geos-fail-bounding-box-search.mp4)

**Comment by lamhuy** [comment]
Current due to this ticket https://codice.atlassian.net/browse/DDF-3857 only AND operation is support by the FilterVisitor
https://github.com/codice/ddf/blob/master/catalog/opensearch/catalog-opensearch-source/src/main/java/org/codice/ddf/opensearch/source/OpenSearchFilterVisitor.java#L401

If you try again with AND, it actually does an OR operation and the record would be found.

**Comment by lamhuy** [comment]
🎉 hero successfully by @BernardIgiri . Please merge to master
